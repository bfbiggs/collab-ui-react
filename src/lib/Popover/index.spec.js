import React from 'react';
import { shallow, mount } from 'enzyme';
import { Popover } from '@collab-ui/react';

describe('tests for <Popover />', () => {
  beforeAll(() => {
    jest.clearAllTimers();
    jest.useFakeTimers();
  });

  it('should match SnapShot', () => {
    const content = <span key="1">Hello how are you doing</span>;
    const container = shallow(
      <Popover content={content}>
        <button>Hello</button>
      </Popover>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render one Popover on click', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover content={content} popoverTrigger={'Click'}>
        <button className="anchor">Hello</button>
      </Popover>
    );

    container.find('.anchor').simulate('click');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('click');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);
  });

  it('focus -> mouseEnter -> mouseLeave -> blur, when popover trigger is MouseEnter', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover content={content} popoverTrigger={'MouseEnter'}>
        <button tabIndex="0" className="anchor">
          Hello
        </button>
      </Popover>
    );

    container.find('.anchor').simulate('focus');
    jest.runAllTimers();
    container.update();
    expect(container.find('.cui-event-overlay__children').length).toEqual(1);

    container.find('.anchor').simulate('mouseenter');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('mouseleave');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('blur');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);
  });

  it('focus -> mouseEnter -> blur -> mouseLeave, when popover trigger is MouseEnter', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover content={content} popoverTrigger={'MouseEnter'}>
        <button tabIndex="0" className="anchor">
          Hello
        </button>
      </Popover>
    );

    container.find('.anchor').simulate('focus');
    jest.runAllTimers();
    container.update();
    expect(container.find('.cui-event-overlay__children').length).toEqual(1);

    container.find('.anchor').simulate('mouseenter');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('blur');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);

    container.find('.anchor').simulate('mouseleave');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);
  });

  it('should render one Popover on mouseenter', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover content={content} popoverTrigger={'MouseEnter'}>
        <button tabIndex="0" className="anchor">
          Hello
        </button>
      </Popover>
    );

    container.find('.anchor').simulate('mouseenter');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('mouseleave');
    jest.runAllTimers();
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);
  });

  it('when show and hide with delay', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover
        content={content}
        popoverTrigger={'MouseEnter'}
        showDelay={200}
        hideDelay={100}
      >
        <button tabIndex="0" className="anchor">
          Hello
        </button>
      </Popover>
    );

    container.find('.anchor').simulate('mouseenter');
    jest.runTimersToTime(300);
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('mouseleave');
    jest.runTimersToTime(200);
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);
  });

  it('when show and hide with delay', () => {
    const content = (
      <span className="popover-content" key="1">
        Hello how are you doing
      </span>
    );
    const container = mount(
      <Popover content={content} popoverTrigger={'MouseEnter'} delay={100}>
        <button tabIndex="0" className="anchor">
          Hello
        </button>
      </Popover>
    );

    container.find('.anchor').simulate('mouseenter');
    jest.runTimersToTime(200);
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    container.find('.anchor').simulate('mouseleave');
    jest.runTimersToTime(99);
    container.update();
    expect(container.find('.popover-content').length).toEqual(1);

    jest.runTimersToTime(200);
    container.update();
    expect(container.find('.popover-content').length).toEqual(0);
  });
});
