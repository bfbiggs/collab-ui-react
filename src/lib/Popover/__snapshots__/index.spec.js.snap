// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <Popover /> should match SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Popover
    className=""
    closeOnClick={false}
    content={
      <span>
        Hello how are you doing
      </span>
    }
    delay={0}
    direction="top-center"
    hideDelay={0}
    popoverTrigger="MouseEnter"
    showArrow={true}
    showDelay={0}
  >
    <button>
      Hello
    </button>
  </Popover>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Hello
        </button>,
        <EventOverlay
          allowClickAway={true}
          anchorNode={null}
          checkOverflow={true}
          className=""
          close={[Function]}
          closeOnClick={false}
          delay={0}
          direction="top-center"
          hideDelay={0}
          isDynamic={false}
          isOpen={false}
          maxHeight={null}
          maxWidth={null}
          showArrow={true}
          showDelay={0}
          style={null}
          targetOffset={
            Object {
              "horizontal": 0,
              "vertical": 0,
            }
          }
        >
          <span>
            Hello how are you doing
          </span>
        </EventOverlay>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "child-1",
        "nodeType": "host",
        "props": Object {
          "children": "Hello",
          "onBlur": [Function],
          "onFocus": [Function],
          "onMouseEnter": [Function],
          "onMouseLeave": [Function],
        },
        "ref": [Function],
        "rendered": "Hello",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowClickAway": true,
          "anchorNode": null,
          "checkOverflow": true,
          "children": <span>
            Hello how are you doing
          </span>,
          "className": "",
          "close": [Function],
          "closeOnClick": false,
          "delay": 0,
          "direction": "top-center",
          "hideDelay": 0,
          "isDynamic": false,
          "isOpen": false,
          "maxHeight": null,
          "maxWidth": null,
          "showArrow": true,
          "showDelay": 0,
          "style": null,
          "targetOffset": Object {
            "horizontal": 0,
            "vertical": 0,
          },
        },
        "ref": [Function],
        "rendered": Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": "Hello how are you doing",
          },
          "ref": null,
          "rendered": "Hello how are you doing",
          "type": "span",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            Hello
          </button>,
          <EventOverlay
            allowClickAway={true}
            anchorNode={null}
            checkOverflow={true}
            className=""
            close={[Function]}
            closeOnClick={false}
            delay={0}
            direction="top-center"
            hideDelay={0}
            isDynamic={false}
            isOpen={false}
            maxHeight={null}
            maxWidth={null}
            showArrow={true}
            showDelay={0}
            style={null}
            targetOffset={
              Object {
                "horizontal": 0,
                "vertical": 0,
              }
            }
          >
            <span>
              Hello how are you doing
            </span>
          </EventOverlay>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "child-1",
          "nodeType": "host",
          "props": Object {
            "children": "Hello",
            "onBlur": [Function],
            "onFocus": [Function],
            "onMouseEnter": [Function],
            "onMouseLeave": [Function],
          },
          "ref": [Function],
          "rendered": "Hello",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowClickAway": true,
            "anchorNode": null,
            "checkOverflow": true,
            "children": <span>
              Hello how are you doing
            </span>,
            "className": "",
            "close": [Function],
            "closeOnClick": false,
            "delay": 0,
            "direction": "top-center",
            "hideDelay": 0,
            "isDynamic": false,
            "isOpen": false,
            "maxHeight": null,
            "maxWidth": null,
            "showArrow": true,
            "showDelay": 0,
            "style": null,
            "targetOffset": Object {
              "horizontal": 0,
              "vertical": 0,
            },
          },
          "ref": [Function],
          "rendered": Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "Hello how are you doing",
            },
            "ref": null,
            "rendered": "Hello how are you doing",
            "type": "span",
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
